pkg/cloudprovider/provider/gce/provider.go

```diff
@@ -0,0 +1,335 @@
+	label := fmt.Sprintf("labels.%s=%s", labelMachineUID, machine.UID)
+	insts, err := svc.Instances.List(cfg.projectID, cfg.zone).Filter(label).Do()
+	if err != nil {
+		if gerr, ok := err.(*googleapi.Error); ok {
+			if gerr.Code == http.StatusNotFound {
+				return nil, errors.ErrInstanceNotFound
+			}
+		}
+		return nil, newError(common.InvalidConfigurationMachineError, errRetrieveInstance, err)
+	}
+	if len(insts.Items) == 0 {
+		return nil, errors.ErrInstanceNotFound
+	}
+	if len(insts.Items) > 1 {
+		return nil, newError(common.InvalidConfigurationMachineError, errGotTooManyInstances)
+	}
+	return &googleInstance{insts.Items[0]}, nil
+}
+
+// GetCloudConfig returns the cloud provider specific cloud-config for the kubelet.
+func (p *Provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	// Read configuration.
+	cfg, err := newConfig(p.resolver, spec.ProviderSpec)
+	if err != nil {
+		return "", "", newError(common.InvalidConfigurationMachineError, errMachineSpec, err)
+	}
+	// Init cloud configuration.
+	cc := &cloudConfig{
+		Global: global{
+			ProjectID: cfg.projectID,
+			LocalZone: cfg.zone,
+		},
+	}
+	config, err = cc.asString()
+	if err != nil {
+		return "", "", newError(common.InvalidConfigurationMachineError, errCloudConfig, err)
+	}
+	return config, "gce", nil
+}
+
+// Create inserts a cloud instance according to the given machine.
+func (p *Provider) Create(
+	machine *v1alpha1.Machine,
+	data *cloud.MachineCreateDeleteData,
+	userdata string,
+) (instance.Instance, error) {
+	// Read configuration.
+	cfg, err := newConfig(p.resolver, machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, errMachineSpec, err)
+	}
+	// Connect to Google compute.
+	svc, err := connectComputeService(cfg)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, errConnect, err)
+	}
+	// Create Google compute instance spec and insert it.
+	networkInterfaces, err := svc.networkInterfaces(cfg)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, errMachineSpec, err)
+	}
+	disks, err := svc.attachedDisks(cfg)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, errMachineSpec, err)
+	}
+	labels := map[string]string{}
+	for k, v := range cfg.labels {
+		labels[k] = v
+	}
+	labels[labelMachineName] = machine.Spec.Name
+	labels[labelMachineUID] = string(machine.UID)
+	inst := &compute.Instance{
+		Name:              machine.Spec.Name,
+		MachineType:       cfg.machineTypeDescriptor(),
+		NetworkInterfaces: networkInterfaces,
+		Disks:             disks,
+		Labels:            labels,
+		ServiceAccounts: []*compute.ServiceAccount{
+			{
+				Email: cfg.jwtConfig.Email,
+				Scopes: []string{
+					compute.ComputeScope,
+					compute.DevstorageReadOnlyScope,
+					"https://www.googleapis.com/auth/monitoring",
+					"https://www.googleapis.com/auth/logging.write",
```

Comment:
Are there any consts for those?

Reply:
Haven't found anywhere, not in `compute` nor elsewhere. But will look again.

Reply:
Maybe you want to temporarily disable the vendor pruning, that helps sometimes finding such things

Reply:
Looks bad, only those few defined here: https://godoc.org/google.golang.org/api/compute/v1#pkg-constants.

Reply:
Found packages for logging with scope constants and `cloud.google.com/go/monitoring/apiv3`. The latter is only exporting `DefaultAuthScopes()` returning a slice of all scopes the package needs.